### PR TITLE
Add more WSLC E2E tests

### PR DESCRIPTION
**Pull request by AmelBawa-msft**
### 😵 WSLCE2EContainerKillTests.cpp‎
- WSLCE2E_Container_Kill_HelpCommand
- WSLCE2E_Container_Kill_KillsRunningContainer
- WSLCE2E_Container_Kill_ByName
- WSLCE2E_Container_Kill_NotFound
- WSLCE2E_Container_Kill_InvalidSignal
- WSLCE2E_Container_Kill_TargetedContainerOnly

### 📋 WSLCE2EContainerListTests.cpp‎
-  WSLCE2E_Container_List_HelpCommand
- WSLCE2E_Container_List_AllOption
- WSLCE2E_Container_List_NoOptions_RunningContainers
- WSLCE2E_Container_List_NoOptions_ExcludesCreatedContainers
- WSLCE2E_Container_List_QuietOption_OutputsIdsOnly
- WSLCE2E_Container_List_InvalidFormatOption
- WSLCE2E_Container_List_JsonFormat

### 🗑️ WSLCE2EContainerRemoveTests.cpp‎
- WSLCE2E_Container_Remove_HelpCommand
- WSLCE2E_Container_Remove_NotFound
- WSLCE2E_Container_Remove_Valid
- WSLCE2E_Container_Remove_ById_Valid
- WSLCE2E_Container_Remove_Force_RunningContainer
- WSLCE2E_Container_Remove_Multiple_Valid

### ⏹️ WSLCE2EContainerStopTests.cpp‎
- WSLCE2E_Container_Stop_HelpCommand
- WSLCE2E_Container_Stop_InvalidSignal
- WSLCE2E_Container_Stop_KillsRunningContainer
- WSLCE2E_Container_Stop_ByName
- WSLCE2E_Container_Stop_NotFound
- WSLCE2E_Container_Stop_TargetedContainerOnly
- WSLCE2E_Container_Stop_SignalByName
- WSLCE2E_Container_Stop_InvalidSignalName
- WSLCE2E_Container_Stop_InvalidTimeout
- WSLCE2E_Container_Stop_ValidTimeoutNegativeOne

### 🖼️ WSLCE2EImageListTests.cpp‎
- WSLCE2E_Image_List_HelpCommand
- WSLCE2E_Image_List_DisplayLoadedImage
- WSLCE2E_Image_List_QuietOption_OutputsNamesOnly
- WSLCE2E_Image_List_InvalidFormatOption
- WSLCE2E_Image_List_JsonFormat

### 🧩 WSLCE2EImageTests.cpp‎
- WSLCE2E_Image_HelpCommand
- WSLCE2E_Image_NoSubcommand_ShowsHelp
- WSLCE2E_Image_InvalidCommand_DisplaysErrorMessage

## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

### CODE REVIEW

File: test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    TEST_METHOD_SETUP(TestMethodSetup)
+    {
+        EnsureContainerDoesNotExist(WslcContainerName);
+        return true;
+    }
+
+    TEST_METHOD(WSLCE2E_Container_Remove_HelpCommand)
+    {
+        auto result = RunWslc(L"container remove --help");
+        result.Verify({.Stdout = GetHelpMessage(), .Stderr = L"", .ExitCode = S_OK});
```
Comment (AmelBawa-msft) [inline test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp]: 0

File: test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    TEST_METHOD(WSLCE2E_Container_Run_Container_With_Command)
+    {
+        VerifyContainerIsNotListed(WslcContainerName);
+
+        auto command = L"echo echo_from_container";
+        auto result = RunWslc(std::format(L"container run --name {} {} {}", WslcContainerName, DebianImage.NameAndTag(), command));
+        result.Verify({.Stdout = L"echo_from_container\n", .Stderr = L"", .ExitCode = S_OK});
+
+        VerifyContainerIsListed(WslcContainerName, L"exited");
+    }
```
Comment (AmelBawa-msft) [inline test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:62]: We have several tests being added by other open PRs:
- #14228 
- #14225 
- #14389 

The ContainerRun tests are being added to the ContainerCreate, but they will be moved to ContainerRun (created by this PR) after they are merged.

File: src/windows/wslc/services/ContainerModel.h [hunk 3]
```diff
@@ -64,6 +64,6 @@ struct ContainerInformation
     ULONGLONG StateChangedAt{};
     ULONGLONG CreatedAt{};
 
-    NLOHMANN_DEFINE_TYPE_INTRUSIVE_ONLY_SERIALIZE(ContainerInformation, Id, Name, Image, State, StateChangedAt, CreatedAt);
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(ContainerInformation, Id, Name, Image, State, StateChangedAt, CreatedAt);
```
Comment (OneBlue) [inline src/windows/wslc/services/ContainerModel.h:69]: not part of this change, but we should use wsl::windows::common::wsla_schema::InspectContainer directly instead of having a copy here

File: test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+namespace WSLCE2ETests {
+
+class WSLCE2EContainerKillTests
+{
+    WSL_TEST_CLASS(WSLCE2EContainerKillTests)
+
+    TEST_CLASS_SETUP(ClassSetup)
+    {
+        EnsureImageIsLoaded(DebianImage);
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp:27]: I think we should rethink this. Re-importing and then deleting this image for every single test class is going to slow down the tests. 

Could we either: 

- Have the CLI test under one class
- Share the images between the different CLI test classes ?

File: test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        VerifyContainerIsListed(firstContainerId, L"running");
+        VerifyContainerIsListed(secondContainerId, L"running");
+
+        // Kill only the first container
+        result = RunWslc(std::format(L"container kill {}", firstContainerId));
+        result.Verify({.Stderr = L"", .ExitCode = 0});
+
+        // Verify first exited, second still running
+        VerifyContainerIsListed(firstContainerId, L"exited");
+        VerifyContainerIsListed(secondContainerId, L"running");
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp:150]: nit: We should also add test coverage for trying to kill a container that's not running

File: test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        auto result = RunWslc(std::format(L"container run -d --name {} {} sleep infinity", WslcContainerName, DebianImage.NameAndTag()));
+        result.Verify({.Stderr = L"", .ExitCode = 0});
+        const auto containerId = result.GetStdoutOneLine();
+        VERIFY_IS_FALSE(containerId.empty());
+
+        VerifyContainerIsListed(containerId, L"running");
+
+        // Removing without force should fail
+        result = RunWslc(std::format(L"container remove {}", containerId));
+        result.Verify({.ExitCode = 1});
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp:121]: We should also validate the error message here

File: test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            result.Verify({.Stderr = L"Invalid signal value: 0 is out of valid range (1-31).\r\n", .ExitCode = 1});
+        }
+
+        {
+            result = RunWslc(std::format(L"container stop {} -s 32 -t 0", WslcContainerName));
+            result.Verify({.Stderr = L"Invalid signal value: 32 is out of valid range (1-31).\r\n", .ExitCode = 1});
+        }
+    }
+
+    TEST_METHOD(WSLCE2E_Container_Stop_KillsRunningContainer)
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp:72]: We should also add test coverage for passing a signal by name (`-s SIGKILL`)

File: test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        VerifyContainerIsListed(firstContainerId, L"running");
+        VerifyContainerIsListed(secondContainerId, L"running");
+
+        // Stop only the first container
+        result = RunWslc(std::format(L"container stop {} -t 0", firstContainerId));
+        result.Verify({.Stderr = L"", .ExitCode = 0});
+
+        // Verify first exited, second still running
+        VerifyContainerIsListed(firstContainerId, L"exited");
+        VerifyContainerIsListed(secondContainerId, L"running");
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp:150]: I recommend adding test coverage for invalid timeout values (invalid integers, negative numbers, etc)

File: test/windows/wslc/e2e/WSLCExecutor.cpp [hunk 9]
```diff
@@ -24,15 +24,28 @@ using namespace wsl::windows::common;
...
+        }
+        ss << L"]";
+        return ss.str();
+    };
+
     Log::Comment((L"Command Line: \"" + CommandLine + L"\"").c_str());
     if (Stdout)
     {
         Log::Comment((L"Stdout: \"" + *Stdout + L"\"").c_str());
+        Log::Comment((L"Stdout (hex): " + DumpHex(*Stdout)).c_str());
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCExecutor.cpp:43]: What's the reason for wanting a hex version of stdout here ? If the objective is to debug CRLF mismatches, I recommend using EscapeString() from Common.h instead

File: src/windows/wslc/services/ContainerModel.h (hunk 3 above)
Comment (AmelBawa-msft) [reply src/windows/wslc/services/ContainerModel.h:69]: Agree, and good catch! I think this was created very early on, before InspectContainer existed, or when the required information had to be gathered from multiple calls

File: test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp (hunk 4 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp:27]: That’s a valid concern, and I agree it could introduce some overhead in the long run. I don’t have a solid solution yet and would prefer to spend more time thinking it through. I also have a few other enhancements in mind that I’d like to tackle once the code and tests are running and stable. I’m tracking those items under issue #14471.

File: test/windows/wslc/e2e/WSLCExecutor.cpp (hunk 9 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCExecutor.cpp:43]: Error messages returned by the service typically use \r\n, whereas those coming from the container (for example, when running it with arguments) usually use \n. The main issue I identified is the presence of hidden whitespace. 

For example:
```
        result.Verify({.Stderr = L"Element not found. \r\nError code: ERROR_NOT_FOUND\r\n", .ExitCode = 1});
                                                      ^ Space hard to capture when copying from the terminal
```

Comment (OneBlue) [reply test/windows/wslc/e2e/WSLCExecutor.cpp:43]: Ok I see. In that case hex is probably OK

File: test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp (hunk 5 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp:150]: Good call! WSLC seems to allow killing a non-running container whereas docker will fail with a descriptive message. I filed a bug to ensure parity. Should the runtime report that error or the CLI would query the state first?

- https://github.com/microsoft/WSL/issues/14509

Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerKillTests.cpp:150]: I will omit this test in this PR until the issue is resolved (potentially in a separate PR)

File: test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp (hunk 6 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp:121]: The error message seems unexpected here. I filed a bug:
- https://github.com/microsoft/WSL/issues/14510

Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerRemoveTests.cpp:121]: This will be addressed in a separate PR

File: test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp (hunk 7 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerStopTests.cpp:72]: Added:
- WSLCE2E_Container_Stop_SignalByName
- WSLCE2E_Container_Stop_InvalidSignalName

File: test/windows/wslc/e2e/WSLCE2EImageListTests.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        return L"The following command aliases are available: ls\r\n\r\n";
+    }
+
+    std::wstring GetAvailableOptions() const
+    {
+        std::wstringstream options;
+        options << L"The following options are available:\r\n"
+                << L"  --format      Output formatting (json or table) (Default:table)\r\n"
+                << L"  -q,--quiet    Outputs the container IDs only\r\n"
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EImageListTests.cpp:151]: - https://github.com/microsoft/WSL/issues/14513

File: test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+    TEST_METHOD(WSLCE2E_Container_Run_Container_With_Command)
+    {
+        WSL2_TEST_ONLY();
+
+        VerifyContainerIsNotListed(WslcContainerName);
+
+        auto command = L"echo echo_from_container";
+        auto result = RunWslc(std::format(L"container run --name {} {} {}", WslcContainerName, DebianImage.NameAndTag(), command));
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:58]: nit: This can be done in a followup change, but we should test the other `run` flags like `-d', `-u`, etc

File: src/windows/wslc/services/ContainerModel.h (hunk 3 above)
Comment (OneBlue) [reply src/windows/wslc/services/ContainerModel.h:69]: Looks like we're still using these in these PR, let's get rid of those in a followup PR ?

File: test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+    TEST_CLASS_CLEANUP(ClassCleanup)
+    {
+        EnsureContainerDoesNotExist(WslcContainerName);
+        return true;
+    }
+
+    TEST_METHOD_SETUP(TestMethodSetup)
+    {
+        EnsureContainerDoesNotExist(WslcContainerName);
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:39]: Doing this for every single test method is going to slow down the tests quite a bit. Instead of using this reset the state of the containers, I recommend using a scope_exit at the method level, [like this for instance](https://github.com/microsoft/WSL/blob/638a3cd571b6c57058b284e7a77fca82899d6555/test/windows/WSLATests.cpp#L446-L452)

Comment (OneBlue) [reply test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:39]: This applies to other places, we should probably rethink the structure a bit, because we're duplicating these for each test class

File: test/windows/wslc/e2e/WSLCExecutor.cpp [hunk 13]
```diff
@@ -25,15 +25,28 @@ using namespace wsl::windows::common;
 
 void WSLCExecutionResult::Dump() const
 {
+    static auto DumpHex = [](const std::wstring& s) {
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCExecutor.cpp:28]: nit: I recommend reusing ` wsl::windows::common::string::BytesToHex()` instead.

**Review by OneBlue — APPROVED** [review]
LGTM, minor comments. Let's think about how we architecture the wslc tests to avoid having  to duplicate the container & image management calls in multiple places in a followup PR

File: test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp (hunk 11 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:58]: Totally agree, lots of new features have been added after this PR was initially created.

I will track this here: https://github.com/microsoft/WSL/issues/14528

I will also experiment with copilot since we have a fairly large number of E2E testing now and testing pattern to follow.

File: test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp (hunk 12 above)
Comment (AmelBawa-msft) [reply test/windows/wslc/e2e/WSLCE2EContainerRunTests.cpp:39]: Fair points, I will think about this, I think we have a couple issues open about E2E enhancements and adding more tests. I am happy to chat later on how we can better align our E2E testing with other parts of the project and potentially meet the project expectations
